[AutoDiff] Support `ref_element_addr` differentiation.

dan-zheng opened the pull request:
Support differentiation of `ref_element_addr`: class stored property references.

Activity analysis:
- Propagate activity for `ref_element_addr` like `struct_element_addr`.
- Do not propagate activity for `ref_element_addr` to `@noDerivative` members.

Pullback generation rules:

```
  Original: y = ref_element_addr x, <n>
   Adjoint: adj[x] += struct (0, ..., #field': adj[y], ..., 0)
                                      ^~~~~~~
                    field in tangent space corresponding to #field
```

- Add `ref_element_addr` case to `PullbackEmitter::getAdjointProjection`.
  - The adjoint projection of a `ref_element_addr` is a local allocation
    initialized with the corresponding field value from the class's base
    adjoint value.

Exposes TF-1149: cannot differentiate active value with loadable type but
address-only `TangentVector` type. Diagnose for now.

Resolves SR-12152.

Todos:
- SR-12152: support class initializer differentiation.
- TF-1149: robust fix for original/tangent value category mismatch.
- TF-970: replace invalid `TangentVector` struct assertions with diagnostics.

---

Example:
```swift
class C: Differentiable {
  @differentiable
  var float: Float

  init(_ float: Float) {
    self.float = float
  }

  @differentiable
  func method(_ x: Float) -> Float {
    x * float
  }
}
print(valueWithGradient(at: C(10), 3, in: { c, x in c.method(x) }))
```
Before:
```console
$ swift tf-645.swift
tf-645.swift:2:4: error: function is not differentiable
  @differentiable
  ~^~~~~~~~~~~~~~
tf-645.swift:3:7: note: when differentiating this function definition
  var float: Float
      ^
tf-645.swift:3:7: note: differentiating class properties is not yet supported
  var float: Float
      ^
```
After:
```
$ swift tf-645.swift
value: 30.0, gradient: (C.TangentVector(float: 3.0), 10.0))
```

dan-zheng committed:
[AutoDiff] Support `ref_element_addr` differentiation.

Support differentiation of `ref_element_addr`: class stored property references.

Teach activity analysis about `ref_element_addr`, handling it like `struct_element_addr`.
Do not propagate activity for `ref_element_addr` to `@noDerivative` members.

Pullback generation rules:
```
Original: y = ref_element_addr x, <n>
 Adjoint: adj[x] += struct (0, ..., #field': adj[y], ..., 0)
                                    ^~~~~~~
                  field in tangent space corresponding to #field
```

Exposes TF-1149: cannot differentiate active value with loadable type but
address-only `TangentVector` type. Diagnose for now.

Resolves SR-12152.

dan-zheng committed:
NFC: fix formatting.

dan-zheng committed:
Add test.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -491,6 +491,10 @@ WARNING(autodiff_nonvaried_result_fixit,none,
         ())
+// TODO(TF-1149): Remove this diagnostic.
+NOTE(autodiff_loadable_value_addressonly_tangent_unsupported,none,
+     "cannot differentiate value with loadable type %0 but address-only "
```
Diagnostics should be user-facing instead of compiler-engineer-facing. How about:

> cannot yet differentiate value with a compile-time known size but a `TangentVector` that contains stored properties of unknown size; consider modifying the `TangentVector` to use fewer generic parameters in stored properties

dan-zheng replied:
Thanks for the tip! Adapted in https://github.com/apple/swift/pull/29749/commits/610214a9c82960e66007edbbf78336a30f310e90:

```
cannot yet differentiate value whose type 'Class<T>' has a compile-time known size, but whose 'TangentVector' contains stored properties of unknown size; consider modifying 'Class<T>.TangentVector' to use fewer generic parameters in stored properties
```

rxwei reviewed (approved):
Very nice!

dan-zheng committed:
Make diagnostic more user-friendly.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Fix incorrect `ref_element_addr` derivative values.

Add `ref_element_addr` case to `PullbackEmitter::getAdjointProjection`.

The adjoint projection of a `ref_element_addr` is a local allocation initialized
with the corresponding field value from the class's base adjoint value.

Fixes incorrect `ref_element_addr` zero derivatives.
Add class initializer differentiation tests.

dan-zheng commented:
https://github.com/apple/swift/pull/29749/commits/5012f1b8cfd475e7ddfc95bc1b4584bf979a88c4 fixes incorrect `ref_element_addr` zero derivatives:

```
Add `ref_element_addr` case to `PullbackEmitter::getAdjointProjection`.

The adjoint projection of a `ref_element_addr` is a local allocation initialized
with the corresponding field value from the class's base adjoint value.
```

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on include/swift/SILOptimizer/Utils/Differentiation/PullbackEmitter.h:
```diff
@@ -354,6 +361,13 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
+  ///    Adjoint: adj[x] += struct (0, ..., #field': adj[y], ..., 0)
+  ///                                       ^~~~~~~
+  ///                     field in tangent space corresponding to #field
+  void visitRefElementAddrInst(RefElementAddrInst *reai);
```
We should get rid of this header soon...

dan-zheng replied:
I can do it soon :)
